distributed/utils_comm.py

```diff
@@ -299,15 +301,20 @@ def subs_multiple(o, d):
 
     """
     typ = type(o)
-    if typ is tuple and o and callable(o[0]):  # istask(o)
+    if not (typ is tuple and o and callable(o[0])):  # not istask(o)
+        try:  # Is o a key to substitute?
+            result = d.get(o, __no_value__)
+            if result is not __no_value__:
+                return result
+        except TypeError:  # unhashable type
+            pass
```

Comment:
We should only try to hash things that aren't tasks, lists of things, or dicts. Perhaps this would be clearer and avoid unnecessary hash attempts?

```python
def subs_multiple(o, d):
    typ = type(o)
    if typ is tuple and o and callable(o[0]):
        return (o[0],) + tuple(fast_subs(i, d) for i in o[1:])
    elif typ is list:
        return [fast_subs(i, d) for i in o]
    elif typ is dict:
        return {k: fast_subs(v, d) for (k, v) in o.items()}
    else:
        try:
            return d.get(o, o)
        except TypeError:
            return o
```